src/main/kotlin/service/impl/TradingServiceImpl.kt

```diff
@@ -0,0 +1,114 @@
+package service.impl
+
+import domain.CryptoExchange
+import domain.User
+import domain.Wallet
+import domain.exception.NotEnoughAmountToExchangeException
+import domain.exception.NotRightPassphraseException
+import domain.exception.SwapTransactionNotProcessException
+import domain.exception.UserNotApprovedException
+import domain.transaction.SwapTransaction
+import domain.transaction.TradeTransaction
+import enums.UserStatus
+import repository.impl.CryptoExchangeRepository
+import repository.impl.TransactionRepository
+import service.TradingService
+import java.math.BigDecimal
+import java.time.Instant
+import domain.Currency
+import kotlin.random.Random
+
+class TradingServiceImpl(
+    private val cryptoExchangeRepository: CryptoExchangeRepository,
+    private val transactionRepository: TransactionRepository
+) : TradingService {
+
+    override fun exchangeCryptoCurrency(
+        firstUser: User,
+        secondUser: User,
+        firstUserCryptoCurrency: Pair<Currency, BigDecimal>,
+        secondUserCryptoCurrency: Pair<Currency, BigDecimal>,
+        cryptoExchange: CryptoExchange
+    ) {
+        validateUserStatus(firstUser)
+        validateUserStatus(secondUser)
+
+        val firstUserWallet = findUserWalletThatContainEnoughCryptoCurrency(
+            user = firstUser,
+            userCryptoCurrency = firstUserCryptoCurrency
+        )
+        val secondUserWallet = findUserWalletThatContainEnoughCryptoCurrency(
+            user = firstUser,
+            userCryptoCurrency = firstUserCryptoCurrency
+        )
+
+        exchangeCurrency(firstUserWallet, firstUserCryptoCurrency, secondUserCryptoCurrency)
+        exchangeCurrency(secondUserWallet, secondUserCryptoCurrency, firstUserCryptoCurrency)
+
+        val tradeTransaction = TradeTransaction(
+            Instant.now(),
+            firstUser,
+            firstUserCryptoCurrency.first,
+            firstUserCryptoCurrency.second,
+            secondUserCryptoCurrency.first,
+            secondUserCryptoCurrency.second
+        )
+
+        cryptoExchange.transactionHistory.add(tradeTransaction)
+        transactionRepository.save(tradeTransaction)
+        cryptoExchangeRepository.save(cryptoExchange)
+    }
+
+    override fun swapCrypto(
+        wallet: Wallet,
+        passphrase: String,
+        cryptoExchange: CryptoExchange,
+        toCurrency: Currency,
+        user: User
+    ) {
+        if (wallet.passphrase != passphrase) {
+            throw NotRightPassphraseException()
+        }
```

Comment:
You can omit {} braces here

Reply:
Should I always omit {} when my if only includes one line? And the same question with loops